new/ohai_identity_attributes.md

```diff
@@ -0,0 +1,124 @@
+---
+RFC: unassigned
+Author: Lamont Granquist <lamont@chef.io>
+Status: Draft
+Type: <Standards Track>
+---
+
+# Ohai Identity Attributes
+
+The purpose of this RFC is to clarify the semantics of the node identity attributes:
+
+* hostname
+* machinename
+* fqdn
+* domain
+
+This RFC encodes the current defacto standards as of Chef 12.1.x / Ohai 8.1.x
+
+## Motivation
+
+    As a Chef user
+    I want consistent semantics for node identity attributes across platforms
+
+    As a Cookbook author
+    I want to be able to use node['fqdn'] and node['domain'] and have them DWIM
+
+    As a Cookbook author
+    I do not want to have to use attributes other than node['fqdn'] or node['domain'] to DWIM
+
+    As a Solaris user
+    I want the node['fqdn'] and node['domain'] to be computed from the short hostname and NIS domain
+
+    As a Linux user
+    I typically want to use the FQDN as my nodename and use that for derived attributes directly
+
+    As a Chef user
+    I do not want DNS failures to affect computation of my nodename/fqdn
+
+## History
+
+In Ohai 7.0.0 the semantics of the identity attributes were normalized to have the following meanings:
+
+* `hostname`:  the short hostname discovered via `hostname` (all platforms were consistent here)
+* `machinename`:  the raw output of `/bin/hostname` (this was introduced so that it was available for node
+  name resolution inside of chef-client)
+* `fqdn`:  equivalent to `/bin/hostname -f` (most platforms used this algorithm before 7.0.0, after 7.0.0 
+  it was made consistent across all platforms -- in some cases this was the return of `/bin/hostname` and
+  Solaris used hostname + NIS domain).
+* `domain`: computed from the domain suffix of the `fqdn` attribute
+
+Changing the `hostname` to be something other than the short hostname is most likely unwise since it has been
+that way consistently since it was introduced.  Because of that the `machinename` attribute was introduced to
+expose the actual return of `/bin/hostname` which is used in `node_name` guessing in the Chef::Client which
+uses the algorithm:
+
+```ruby
+name = Chef::Config[:node_name] || ohai[:fqdn] || ohai[:machinename] || ohai[:hostname]
+```
+
+It was discussed making the `fqdn` attribute to prefer the return of `/bin/hostname` when that returned a
+FQDN, but this felt to break some edge conditions in DNS related cookbooks (djbdns?) and so this was forced
+to be the fqdn which was resolved via DNS resolution across all platforms.  For platforms that do not support
+the `-f` flag to `hostname` a pure ruby version is used by ohai.
+
+The changes in Ohai 7.0.0 to the `fqdn` attribute broke Solaris people with busted DNS who expected the
```

Comment:
I believe they do.  Broken is broken, though.  I can be more verbose and talk around it, but if people didn't have broken DNS this wouldn't be an issue since the forward+reverse lookup would not fail.